taichi/common/serialization.h

```diff
@@ -54,71 +56,93 @@ using is_unit = typename std::is_base_of<Unit, remove_cvref_t<T>>;
 
 template <typename T>
 using is_unit_t = typename is_unit<T>::type;
+
 }  // namespace type
 
-#define TI_IO_DECL_INST                               \
-  void binary_io(BinaryOutputSerializer &ser) const { \
-    ser(*this);                                       \
-  }                                                   \
-  void binary_io(BinaryInputSerializer &ser) const {  \
-    ser(*this);                                       \
-  }
+namespace detail {
 
-#define TI_IO_DECL_INST_VIRT                                  \
-  virtual void binary_io(BinaryOutputSerializer &ser) const { \
-    ser(*this);                                               \
-  }                                                           \
-  virtual void binary_io(BinaryInputSerializer &ser) const {  \
-    ser(*this);                                               \
+template <size_t N>
+constexpr size_t count_delim(const char (&str)[N], char delim) {
+  size_t count = 1;
+  for (const char &ch : str) {
+    if (ch == delim) {
+      ++count;
+    }
   }
+  return count;
+}
 
-#define TI_IO_DECL_INST_VIRT_OVERRIDE                                  \
-  virtual void binary_io(BinaryOutputSerializer &ser) const override { \
-    ser(*this);                                                        \
-  }                                                                    \
-  virtual void binary_io(BinaryInputSerializer &ser) const override {  \
-    ser(*this);                                                        \
+template <size_t DelimN>
+struct StrDelimSplitter {
+  template <size_t StrsN>
+  static constexpr std::array<std::string_view, DelimN> make(
+      const char (&str)[StrsN],
+      char delim) {
+    std::array<std::string_view, DelimN> res;
+    const char *head = &(str[0]);
+    size_t si = 0;
+    size_t cur_head_i = 0;
+    size_t ri = 0;
+    while (si < StrsN) {
+      if (str[si] != delim) {
+        si += 1;
+      } else {
+        res[ri] = {head, (si - cur_head_i)};
+        ++ri;
+        si += 2;  // skip ", "
```

Comment:
What if there is no " " after ","? Can we write something like `.strip()` in Python here?

Reply:
This is mostly handling the string created via macros (e.g., `#__VA_ARGS__`), so it should be fine.

Reply:
I see!